File: development/skills/bootstrap/templates/languages/spring/resilience/DependencyHealth.java
/*
 * Passive dependency health, read from circuit-breaker state (#1141, epic #964).
 *
 * The unifying idea of the org resilience policy: THE CIRCUIT BREAKER KEEPS YOU
 * SERVING; THE DEPENDENCY-HEALTH SURFACE TELLS YOU WHAT'S DEGRADED. An open
 * breaker IS a down dependency, so health is read from the breaker rather than
 * measured by a second mechanism.
 *
 * PASSIVE means exactly that: this class never calls a dependency, never runs a
 * scheduled probe, and NEVER transitively calls a downstream's /health (the
 * health-check-storm anti-pattern, where one slow leaf hangs every ancestor's
 * health check). Real request traffic -- or the breaker's own half-open probe --
 * has already moved the state; reading it costs nothing and generates no traffic.
 *
 * PLACEMENT: copy into your service's source set and set the package to match its
 * directory, e.g. src/main/java/com/acme/orders/ops/DependencyHealth.java ->
 * package com.acme.orders.ops;
 */
package com.example.ops; // <-- CHANGE to match the directory you place this file in.

import io.github.resilience4j.circuitbreaker.CircuitBreaker;
import io.github.resilience4j.circuitbreaker.CircuitBreakerRegistry;
import java.time.Clock;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.boot.availability.ApplicationAvailability;
import org.springframework.boot.availability.ReadinessState;
import org.springframework.stereotype.Component;

/**
 * Derives the ops-api v1.1 dependency-health view from resilience4j breaker state.
 *
 * <p>The mapping from breaker state to dependency status is exact and defined by the
 * contract: closed = {@code up}, half-open = {@code degraded} (being re-probed), open =
 * {@code down}.
 *
 * <p>An ordinary component of the MAIN application context: it serves no HTTP itself, so
 * scanning it here is safe, and OpsHealthEndpoint (the Actuator endpoint that does serve
 * it, on the management port) injects it normally.
 */
@Component
public class DependencyHealth {

  /**
   * One direct dependency's health, serialized as an ops-api {@code DependencyHealth}.
   *
   * @param status up | degraded | down -- read from the breaker
   * @param kind hard | soft -- the declared readiness hinge
   * @param breaker closed | open | half_open -- the state the status was read from
   * @param since RFC 3339 timestamp of the last state change, so a dashboard can tell a
   *     blip from a sustained outage
   */
  public record Component(String status, String kind, String breaker, String since) {}

  private final CircuitBreakerRegistry registry;
  private final DependencyCatalog catalog;
  private final ApplicationAvailability availability;
  private final Clock clock;

  /** Dependency name -> when it last changed breaker state. */
  private final Map<String, Instant> since = new ConcurrentHashMap<>();

  /**
   * The Clock is taken from the context when the service defines one and defaults to
   * {@link Clock#systemUTC()} otherwise, so this never competes with an application's own
   * Clock bean while staying injectable in tests.
   */
  public DependencyHealth(
      CircuitBreakerRegistry registry,
      DependencyCatalog catalog,
      ApplicationAvailability availability,
      ObjectProvider<Clock> clock) {
    this.registry = registry;
    this.catalog = catalog;
    this.availability = availability;
    this.clock = clock.getIfAvailable(Clock::systemUTC);
    subscribe();
  }

  /**
   * Bind each declared dependency to its breaker and start tracking state changes.
   *
   * <p>A declared dependency with no breaker is a HARD startup failure, not a warning.
   * Mandate 2 says every governed outbound call has a breaker; without one this class
   * would have nothing to read and would report the dependency as {@code up}
   * unconditionally -- precisely the under-reporting the ops-api contract forbids. Fail
   * loudly at boot instead of serving a health surface that lies during an outage.
   */
  private void subscribe() {
    Instant startedAt = now();
    catalog
        .resolved()
        .forEach(
            (name, dep) -> {
              CircuitBreaker breaker =
                  registry
                      .find(dep.breaker())
                      .orElseThrow(
                          () ->
                              new IllegalStateException(
                                  "Dependency '"
                                      + name
                                      + "' is declared under resilience.dependencies but has no"
                                      + " resilience4j circuit breaker named '"
                                      + dep.breaker()
                                      + "'. Declare it under"
                                      + " resilience4j.circuitbreaker.instances, or drop the"
                                      + " dependency declaration."));
              since.put(name, startedAt);
              breaker.getEventPublisher().onStateTransition(event -> since.put(name, now()));
            });
  }

  /**
   * The ops-api {@code components} map: one entry per DIRECT dependency, in declaration
   * order. Empty when the service declares no dependencies, in which case the caller
   * omits the field entirely.
   */
  public Map<String, Component> components() {
    Map<String, Component> out = new LinkedHashMap<>();
    catalog
        .resolved()
        .forEach(
            (name, dep) -> {
              CircuitBreaker.State state = registry.circuitBreaker(dep.breaker()).getState();
              out.put(
                  name,
                  new Component(
                      statusOf(state),
                      dep.kind().name().toLowerCase(java.util.Locale.ROOT),
                      breakerOf(state),
                      rfc3339(since.getOrDefault(name, now()))));
            });
    return out;
  }

  /**
   * The aggregate served by {@code /health}.
   *
   * <p>The components set a FLOOR, not an equality: {@code down} if any HARD dependency
   * is down, else {@code degraded} if ANY dependency (hard or soft) is down or degraded,
   * else {@code ok}. Note the hard/degraded case -- a hard dependency that is merely
   * half-open is {@code degraded}, not down.
   *
   * <p>Over-reporting is legal and expected: a service that is impaired for an internal
   * reason no dependency models (a backed-up work queue, a full disk) MUST report a MORE
   * severe aggregate than this. Add that here; never report a LESS severe one.
   */
  public String aggregate() {
    String floor = "ok";
    for (Map.Entry<String, Component> entry : components().entrySet()) {
      Component component = entry.getValue();
      boolean down = "down".equals(component.status());
      if (down && "hard".equals(component.kind())) {
        return "down";
      }
      if (down || "degraded".equals(component.status())) {
        floor = "degraded";
      }
    }
    return floor;
  }

  /**
   * The readiness answer behind {@code /health/ready}.
   *
   * <p>Among dependencies, only a declared HARD one being down makes the service unready
   * -- a soft dependency down keeps the pod ready and serving degraded responses. That
   * single rule resolves the tension between naive readiness (shed traffic on any
   * dependency loss) and resilience (stay up and degrade).
   *
   * <p>Non-dependency reasons are unaffected and still apply: a service still starting
   * up, or draining during a graceful shutdown, is unready even with every dependency
   * healthy. That state comes from Spring's own {@link ReadinessState}.
   */
  public boolean ready() {
    if (availability.getReadinessState() != ReadinessState.ACCEPTING_TRAFFIC) {
      return false;
    }
    return components().values().stream()
        .noneMatch(c -> "hard".equals(c.kind()) && "down".equals(c.status()));
  }

  /**
   * Breaker state -> dependency status.
   *
   * <p>FORCED_OPEN is an operator-forced open breaker: calls are being rejected, so the
   * dependency is unusable and reports {@code down} exactly like a naturally open one.
   * DISABLED and METRICS_ONLY mean the breaker is not rejecting anything, so it has no
   * failure verdict to report and the dependency reads {@code up}.
   */
  private static String statusOf(CircuitBreaker.State state) {
    return switch (state) {
      case OPEN, FORCED_OPEN -> "down";
      case HALF_OPEN -> "degraded";
      case CLOSED, DISABLED, METRICS_ONLY -> "up";
    };
  }

  /**
   * Breaker state -> the contract's three-value {@code breaker} field. resilience4j has
   * six states; the contract names the three that describe what is happening to calls, so
   * the non-enforcing states project onto {@code closed} and FORCED_OPEN onto {@code
   * open}.
   */
  private static String breakerOf(CircuitBreaker.State state) {
    return switch (state) {
      case OPEN, FORCED_OPEN -> "open";
      case HALF_OPEN -> "half_open";
      case CLOSED, DISABLED, METRICS_ONLY -> "closed";
    };
  }

  private Instant now() {
    return clock.instant().truncatedTo(ChronoUnit.MILLIS);
  }

  private static String rfc3339(Instant instant) {
    return instant.truncatedTo(ChronoUnit.MILLIS).toString();
  }
}

File: development/skills/bootstrap/templates/languages/python/resilience/dependency_health.py
"""Passive dependency health, read from circuit-breaker state, for Python (#1143, epic #964).

The unifying idea of the org resilience policy: THE CIRCUIT BREAKER KEEPS YOU SERVING;
THE DEPENDENCY-HEALTH SURFACE TELLS YOU WHAT'S DEGRADED. An open breaker IS a down
dependency, so health is read from the breaker rather than measured by a second mechanism.

PASSIVE means exactly that: this module never calls a dependency, never runs a scheduled
probe, and NEVER transitively calls a downstream's /health (the health-check-storm
anti-pattern, where one slow leaf hangs every ancestor's health check). Real request
traffic has already moved the state; reading it costs nothing and generates no traffic.

WHERE THE AGGREGATE AND THE READINESS RULE LIVE: not here. This module maps breaker state
to the per-dependency ``components`` entries and stops there; ``ops_api.OpsConfig`` derives
the aggregate floor and the readiness answer from those entries.

IT IMPORTS NO BREAKER LIBRARY, deliberately. It needs one thing from a breaker -- its
``state`` string -- so it reads that attribute and nothing else. Two things follow: the ops
surface can be unit-tested with a two-line fake breaker, and swapping the library (should
this family ever re-bless one) touches ``dependency_catalog`` alone.

PLACEMENT: copy into your service's package BESIDE ``ops_api.py`` and
``dependency_catalog.py``.
"""

from __future__ import annotations

from collections.abc import Mapping
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Callable

# PLACEMENT-PROOF IMPORT, and it is not boilerplate. The documented placement puts this
# module inside your package (``src/<pkg>/``), where Python 3 has no implicit relative
# imports: a bare ``from ops_api import Dependency`` then raises ModuleNotFoundError at
# startup, because ``sys.path`` carries ``src/``, not ``src/<pkg>/``. The relative form is
# therefore FIRST. The fallback covers the flat layout (files in one directory, run as
# top-level modules), which is how the payload's own tests exercise it.
try:
    from .ops_api import Dependency
except ImportError:  # not a package -- flat layout
    from ops_api import Dependency

if TYPE_CHECKING:  # pragma: no cover - import-time only, keeps this module library-free
    try:
        from .dependency_catalog import DependencyCatalog
    except ImportError:
        from dependency_catalog import DependencyCatalog

#: Breaker state -> the dependency status the contract reports.
#:
#: The mapping is exact and defined by the contract: closed = ``up``, half-open =
#: ``degraded`` (being re-probed), open = ``down``. ``circuitbreaker`` spells its three
#: states exactly as the contract's ``breaker`` field does, so the ``breaker`` value needs
#: no translation -- but map through this table rather than passing the raw string on, so a
#: library that ever spelled a state differently fails loudly here instead of emitting an
#: off-contract value the conformance checker rejects at deploy time.
_STATUS_BY_STATE = {
    "closed": "up",
    "half_open": "degraded",
    "open": "down",
}


class DependencyHealth:
    """Derives the ops-api v1.1 ``components`` map from circuit-breaker state.

    Wire it into the ops surface at startup::

        catalog = DependencyCatalog.load()
        ...build your dependency clients, each calling catalog.require_declared(...)...
        catalog.require_all_declared_guarded()
        serve(config=OpsConfig(dependencies=DependencyHealth(catalog)))
    """

    def __init__(self, catalog: DependencyCatalog, now: Callable[[], datetime] | None = None) -> None:
        """The ``now`` callable is injectable so a test can assert on ``since`` without sleeping."""
        self._catalog = catalog
        self._now = now or (lambda: datetime.now(timezone.utc))
        started_at = self._rfc3339(self._now())
        # name -> (last observed state, when this service first observed it)
        self._observed: dict[str, tuple[str, str]] = {
            name: (self._state_of(name), started_at) for name in catalog.dependencies
        }

    def components(self) -> Mapping[str, Dependency]:
        """One entry per DIRECT dependency, in declaration order.

        Empty when the service declares none, in which case ``ops_api`` omits the field
        entirely and the response stays a valid ops-api v1.0 body.
        """
        out: dict[str, Dependency] = {}
        for name, kind in self._catalog.dependencies.items():
            state = self._state_of(name)
            out[name] = Dependency(
                status=self._status_of(state, name),
                kind=kind,
                breaker=state,
                since=self._since(name, state),
            )
        return out

    def _since(self, name: str, state: str) -> str:
        """When this service first OBSERVED the dependency in its current state (RFC 3339).

        Observation-stamped, and the wording is the honest one. ``circuitbreaker`` publishes
        no state-change callback, so there is no transition to hook; what this can truthfully
        report is when the change was first SEEN. Every /health scrape is an observation, so
        the stamp lands within one scrape interval of the real transition. The alternative -- deriving it from the breaker's internals --
        would read a private attribute, and would still have nothing to offer for the closed
        state, which has no timestamp at all.
        """
        previous_state, previous_since = self._observed.get(name, (state, ""))
        if previous_state != state or not previous_since:
            stamp = self._rfc3339(self._now())
            self._observed[name] = (state, stamp)
            return stamp
        return previous_since

    def _state_of(self, name: str) -> str:
        """The breaker's own word for its state.

        Read as a plain attribute, with no call and no lock: ``circuitbreaker`` computes the
        open -> half-open transition from elapsed time inside the property itself. That is
        what makes MANDATE 5 (background reconnect) observable with no traffic and no deploy
        -- a recovering dependency shows up as ``degraded`` on the next scrape rather than
        staying ``down`` until something happens to call it.
        """
        return str(self._catalog.breaker(name).state)

    @staticmethod
    def _status_of(state: str, name: str) -> str:
        try:
            return _STATUS_BY_STATE[state]
        except KeyError:  # pragma: no cover - only reachable if the breaker library changes
            raise ValueError(
                f"dependency {name!r} reported unknown breaker state {state!r}; "
                f"expected one of {sorted(_STATUS_BY_STATE)}"
            ) from None

    @staticmethod
    def _rfc3339(moment: datetime) -> str:
        """RFC 3339 with a 'Z' offset and no sub-second noise.

        ``datetime.isoformat()`` renders UTC as ``+00:00``; both spell the same instant and
        the contract's ``date-time`` format accepts either, but ``Z`` is what every sibling
        payload emits and a dashboard grouping on the raw string should not see two forms.
        """
        return moment.astimezone(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")
